hal-x86_64/src/interrupt/mod.rs

```diff
@@ -0,0 +1,419 @@
+pub mod idt;
+pub mod pic;
+use crate::{cpu, segment, vga, VAddr};
+use core::{
+    fmt::{self, Write},
+    marker::PhantomData,
+};
+pub use idt::Idt;
+pub use pic::CascadedPic;
+
+use hal_core::interrupt::{ctx, Handlers};
+
+#[derive(Debug)]
+#[repr(C)]
+pub struct Context<'a, T = ()> {
+    registers: &'a mut Registers,
+    code: T,
+}
+
+pub type ErrorCode = u64;
+
+/// An interrupt service routine.
+pub type Isr<T> = extern "x86-interrupt" fn(&mut Context<T>);
+
+#[derive(Debug)]
+#[repr(C)]
+pub struct Interrupt<T = ()> {
+    vector: u8,
+    _t: PhantomData<T>,
+}
+
+#[derive(Copy, Clone)]
+#[repr(transparent)]
+pub struct PageFaultCode(u64);
+
+#[repr(C)]
+pub struct Registers {
+    pub instruction_ptr: VAddr, // TODO(eliza): add VAddr
+    pub code_segment: segment::Selector,
+    _pad: [u16; 3],
+    pub cpu_flags: u64,   // TODO(eliza): rflags type?
+    pub stack_ptr: VAddr, // TODO(eliza): add VAddr
+    pub stack_segment: segment::Selector,
+    _pad2: [u16; 3],
+}
+
+static mut IDT: idt::Idt = idt::Idt::new();
+static mut PIC: pic::CascadedPic = pic::CascadedPic::new();
+
+#[cold]
+#[inline(never)]
+extern "x86-interrupt" fn nop(cx: Context<'_>) {}
+
+static mut TIMER: usize = 0;
+
+struct TestHandlersImpl;
+
+impl Handlers<crate::X64> for TestHandlersImpl {
+    #[inline(never)]
```

Comment:
oh why are these `inline(never)`? i'd have expected they'd be `inline(always)` even, so they get inlined into trampolines

Reply:
they should be `#[inline(always)]` (and actually, i think that without an inline attr, rustc will probably inline them anyway since they never get called elsewhere...); i added this attr while debugging the issue that turned out to be caused by us not remapping PICs.